### PR TITLE
BUG: GH42866 DatetimeIndex de-serializing fails in PYTHONOPTIMIZE mode

### LINKED ISSUE #42866: BUG: DatetimeIndex de-serializing fails in PYTHONOPTIMIZE mode (python -O)
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

Run python as PYTHONOPTIMIZE, option -O:

```bash
python -O
```

```python
import pickle
import pandas as pd

idx = pd.date_range('2021-01-01', '2021-01-02', freq='D')
idx_dump = pickle.dumps(idx)
pickle.loads(idx_dump)

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../lib/python3.7/site-packages/pandas/core/indexes/datetimes.py", line 61, in _new_DatetimeIndex
    result = cls._simple_new(dta, **d)
TypeError: _simple_new() got an unexpected keyword argument 'tz'

```

#### Problem description

These dict operations go under the 'assert' statement:

```python

# pandas/core/indexes/datetimes.py

def _new_DatetimeIndex(cls, d):
    ...
        if key in d:
            assert d.pop(key) == getattr(dta, key)

    ...

```

which are ignored in PYTHONOPTIMIZED mode.
As a fix, bring dict pop operation out of the assert statement.

#### Expected Output

```python

DatetimeIndex(['2021-01-01', '2021-01-02'], dtype='datetime64[ns]', freq='D')

```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 7c48ff4409c622c582c56a5702373f726de08e96
python           : 3.7.9.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.5.0
Version          : Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.2.5
numpy            : 1.20.3
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.1.2
setuptools       : 57.0.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : None
psycopg2         : 2.8.6 (dt dec pq3 ext lo64)
jinja2           : None
IPython          : 7.20.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.3
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by GorAlexander**
- [x] closes #42866
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
Thanks @GorAlexander 

@jbrockmendel does pandas have any way of testing code run in `-O` mode?

**Comment by jbrockmendel** [comment]
> does pandas have any way of testing code run in -O mode?

take a look at test_oo_optimizable

**Comment by MarcoGorelli** [comment]
Thanks - @GorAlexander do you want to have a look there and add a test?

**Comment by jreback** [comment]
looks fine, can you add a whatsnew note. 1.4 bug fix, Datetimelike section. ping on green.

**Comment by GorAlexander** [comment]
@github-actions pre-commit

**Comment by GorAlexander** [comment]
Hi @jreback , I have fixed code-styles and added whatsnew note.
Is it possible to apply this fix to version 1.3?

**Comment by jreback** [comment]
would be ok for 1.3.x if u can move the note

File: doc/source/whatsnew/v1.4.0.rst [hunk 1]
```diff
@@ -193,6 +193,7 @@ Categorical
 Datetimelike
 ^^^^^^^^^^^^
 - Bug in :class:`DataFrame` constructor unnecessarily copying non-datetimelike 2D object arrays (:issue:`39272`)
+- Bug in :func:`_new_DatetimeIndex` when de-serializing datetime indexes in PYTHONOPTIMIZED mode (:issue:`42866`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:196]: ok to move to 1,3.2

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.4.0.rst (hunk 1 above)
Comment (GorAlexander) [reply doc/source/whatsnew/v1.4.0.rst:196]: @jreback ok, so I'll have to fork from 1.3.x branch and open a new PR

**Comment by jreback** [comment]
no we merge to master and then backport

just change the note

**Comment by GorAlexander** [comment]
@jreback , thanks! I've moved the note to v1.3.2, so it must be ready.

**Comment by GorAlexander** [comment]
Hi @jreback , the whatsnew note was corrected, however GH still highlights about 'changes requested'. Is there smth else to do to get the PR approved?

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me

Tried this out and it fails on `master` but passes here 👍 
```console
(pandas-dev) marcogorelli@OVMG025 pandas-dev % pytest pandas/tests/test_downstream.py -k test_oo_optimized_datetime_index_unpickle -qq
.
------- generated xml file: /Users/marcogorelli/pandas-dev/test-data.xml -------
============================= slowest 30 durations =============================
0.55s call     pandas/tests/test_downstream.py::test_oo_optimized_datetime_index_unpickle

(2 durations < 0.005s hidden.  Use -vv to show these durations.)
(pandas-dev) marcogorelli@OVMG025 pandas-dev % git checkout upstream/master -- pandas/core/indexes/datetimes.py                    
(pandas-dev) marcogorelli@OVMG025 pandas-dev % pytest pandas/tests/test_downstream.py -k test_oo_optimized_datetime_index_unpickle -qq
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/marcogorelli/pandas-dev/pandas/core/indexes/datetimes.py", line 96, in _new_DatetimeIndex
    result = cls._simple_new(dta, **d)
TypeError: _simple_new() got an unexpected keyword argument 'freq'
F
================================================ FAILURES =================================================
________________________________ test_oo_optimized_datetime_index_unpickle ________________________________

    def test_oo_optimized_datetime_index_unpickle():
        # GH 42866
>       subprocess.check_call(
            [
                sys.executable,
                "-OO",
                "-c",
                (
                    "import pandas as pd, pickle; "
                    "pickle.loads(pickle.dumps(pd.date_range('2021-01-01', periods=1)))"
                ),
            ]
        )

pandas/tests/test_downstream.py:78: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

popenargs = (['/Users/marcogorelli/opt/miniconda3/envs/pandas-dev/bin/python', '-OO', '-c', "import pandas as pd, pickle; pickle.loads(pickle.dumps(pd.date_range('2021-01-01', periods=1)))"],)
kwargs = {}, retcode = 1
cmd = ['/Users/marcogorelli/opt/miniconda3/envs/pandas-dev/bin/python', '-OO', '-c', "import pandas as pd, pickle; pickle.loads(pickle.dumps(pd.date_range('2021-01-01', periods=1)))"]

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.
    
        The arguments are the same as for the call function.  Example:
    
        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '['/Users/marcogorelli/opt/miniconda3/envs/pandas-dev/bin/python', '-OO', '-c', "import pandas as pd, pickle; pickle.loads(pickle.dumps(pd.date_range('2021-01-01', periods=1)))"]' returned non-zero exit status 1.

../opt/miniconda3/envs/pandas-dev/lib/python3.8/subprocess.py:364: CalledProcessError
-------------------- generated xml file: /Users/marcogorelli/pandas-dev/test-data.xml ---------------------
========================================== slowest 30 durations ===========================================
0.58s call     pandas/tests/test_downstream.py::test_oo_optimized_datetime_index_unpickle

(2 durations < 0.005s hidden.  Use -vv to show these durations.)
========================================= short test summary info =========================================
FAILED pandas/tests/test_downstream.py::test_oo_optimized_datetime_index_unpickle - subprocess.CalledPro...
(pandas-dev) marcogorelli@OVMG025 pandas-dev % 
```

File: doc/source/whatsnew/v1.3.2.rst [hunk 2]
```diff
@@ -32,6 +32,8 @@ Bug fixes
 ~~~~~~~~~
 - 1D slices over extension types turn into N-dimensional slices over ExtensionArrays (:issue:`42430`)
 - :meth:`.Styler.hide_columns` now hides the index name header row as well as column headers (:issue:`42101`)
+- Bug in :func:`_new_DatetimeIndex` when de-serializing datetime indexes in PYTHONOPTIMIZED mode (:issue:`42866`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.2.rst:35]: this won't render and is an internal function, so just
```suggestion
- Bug in de-serializing datetime indexes in PYTHONOPTIMIZED mode (:issue:`42866`)
```

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @GorAlexander

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**